Make structures in loki connector public

vyzigold opened the pull request:
This allows a programmer to create it's own constructors
outside of apputils. My current usecase for this is a
custom constructor in sg-core, that uses a different config.

paramite commented:
So you could use some struct for in/out channels, but the reason why interface{} has been used in existing connectors is to be able to create interface for connectors so that in theory user can then switch connectors as plugins with minimal or no changes in application implementation. You can use type switch [1] on the channel data and basically use any object in those channels.

[1] https://tour.golang.org/methods/16

paramite commented:
For solving streams ... you could have a stream buffer and a configuration option .. something like StreamsBufferSize int. And when the buffer would be full, it will be sent to Loki. That way you can have 1 and n size save with one goroutine (and one method).

paramite commented:
Regarding the rsyslog. It is expected to receive single log record with set of labels with each AMQP message.

vyzigold commented:
> For solving streams ... you could have a stream buffer and a configuration option .. something like StreamsBufferSize int. And when the buffer would be full, it will be sent to Loki. That way you can have 1 and n size save with one goroutine (and one method).

I already have something like that. There is a  "batch size" configuration option, which determines how many streams to send at once.


> So you could use some struct for in/out channels, but the reason why interface{} has been used in existing connectors is to be able to create interface for connectors so that in theory user can then switch connectors as plugins with minimal or no changes in application implementation. You can use type switch [1] on the channel data and basically use any object in those channels.
> 
> [1] https://tour.golang.org/methods/16

I thought about it right after I wrote this, so I'm now working on exactly this.



> Regarding the rsyslog. It is expected to receive single log record with set of labels with each AMQP message. But IMO the set of labels can be completely same for some log records

So it's probably not worth to care about the amount of logs in a single stream, Loki will buffer the streams and sort them on its own and probably with much better performance.

vyzigold commented on connector/loki.go:
```diff
@@ -48,68 +73,127 @@ type endpoints struct {
 // the current batch > maxBatch or if more time
 // than maxWaitTime passed
-func (client *LokiClient) Start() {
+func (client *LokiConnector) Start(outchan chan interface(), inchan chan interface()) {
```
I'm not sure I understand. The inchan (second parameter) is used to take data from user of the connector and send them to loki, outchan would be used for data going from loki to the user, but I decided I won't use outchan. When I look at the amqp connector, it looks the same: inchan (second parameter) is the channel used for data going from connector user to the bus and outchan for data from the bus to the user.

paramite replied:
O_o I could swear I read the channel usage the opposite way this morning, but you're right. Just ignore my previous comment.

paramite commented on config/json.go:
```diff
@@ -172,6 +172,9 @@ func (conf JSONConfig) GetOption(name string) (*Option, error) {
 				option, err = extractValue(opt, optAddr[1:])
 			}
+			if option == nil {
+				err = fmt.Errorf("did not find option '%s' in section '%s'", nameparts[1], nameparts[0])
```
This is not true for structured options in JSONConfig. You can have defined option for parsing, but not found in config, hence the return value is nil.

vyzigold replied:
So the bug is in amqp connector then. It panics when calling `NewAMQP10Connector()` without `ListenPrefetch` specified in the config. Because here:
```
		if listen, err := conf.GetOption("Amqp1.Connection.ListenChannels"); err == nil {
			prefetch := int64(-1)
			prf, err := conf.GetOption("Amqp1.Connection.ListenPrefetch")
			if err == nil {
				prefetch = prf.GetInt()
			}
			for _, channel := range listen.GetStrings(",") {
				connector.CreateReceiver(channel, int(prefetch))
			}
		}

```
prf is nil when calling prf.GetInt().

vyzigold replied:
But I guess I'll just remove that commit, because it's a completely different issue and it probably doesn't belong in this PR.

paramite commented on connector/loki.go:
```diff
@@ -48,43 +73,83 @@ type endpoints struct {
+		} else {
+			return nil, fmt.Errorf("Failed to get connection URL from configuration file at loki/connection")
+		}
+		if batchSize, err := conf.GetOption("loki/batchSize"); err == nil {
```
Just a nit, but please let's keep consistency and use "batch_size" naming for INIConfig. Similarly for other options

vyzigold committed:
Make structures in loki connector public

This allows a programmer to create it's own constructors
outside of apputils. My current usecase for this is a
custom constructor in sg-core, that uses a different config.

vyzigold commented:
Martin had a better idea, I'll take another look at it tomorrow.
